### ISSUE TITLE
Automatically run overdue snapshots

### ISSUE THREAD

**Issue by DarkArc** [kopia/kopia#3319]
https://github.com/kopia/kopia/pull/3311 (implementing https://github.com/kopia/kopia/issues/3197) doesn't seem to work.

![image](https://github.com/kopia/kopia/assets/778012/b9573733-192b-4c26-b296-388f3c44746a)

I set this all up last night, shortly after midnight, shutdown the computer, and then booted it this morning. The snapshot did not run.

**Comment by ghaushe-ampere** [comment]
Please take a screenshot showing this snapshot on the snapshot screen

A snapshot is considered skipped only if the last snapshot was > 24 hours ago.  And the previous snapshot date is set to the last time the policy was updated if there is no snapshot yet (as far as I know).

So if you create a policy, it won't be considered stale for 24 hours.

This should work properly for time-of-day snapshots (since they run once per day) but will not work properly for crontab snapshots that can run more frequently than that.  I'll need to rework the code to properly handle the crontab case, but I think this explains why your snapshot didn't run today.

**Comment by DarkArc** [comment]
Ah, yes that seems to track:
![image](https://github.com/kopia/kopia/assets/778012/5e45ec84-87e1-4db4-bd94-a9378fd1dfdb)

RE: the logic for determining whether a snapshot was skipped; why not have Kopia record the time of the next snapshot when it runs a snapshot (or its configuration is updated -- e.g., if I ran one now it would record "the next snapshot is 1:00AM Eastern Time on Sept 19th, 2023".), and then when Kopia starts up, if it's past that time, it runs?

That seems like a fairly simple design that works for any scheduling pattern.

**Comment by lupusA** [comment]
Hi, 

what about multiple times-of-day snapshots? You can enter multiple times manually, which would basically result in a individual frequency. This needs to be handled accordingly. 

Cheers,

**Comment by PhracturedBlue** [comment]
I think the algorithm in  #3323 should be much more robust.

**Comment by PhracturedBlue** [comment]
I also created a PR to change the default to enable run-missed here:  #3330

**Comment by DarkArc** [comment]
@PhracturedBlue thanks

**Comment by TheRotag** [comment]
Commenting per a forum post I made ([Daily snapshot finishes, shows 0B, retries](https://kopia.discourse.group/t/daily-snapshot-finishes-shows-0b-retries/2194)) where another user suggested I chime in here regarding my issue ([direct comment link](https://kopia.discourse.group/t/daily-snapshot-finishes-shows-0b-retries/2194/12?u=therotag)).

Tl;dr: I have a policy that snapshots at a fixed time every day and takes (on average) about 2-3 hours to run, resulting in multiple "incomplete" snapshots getting created, which are then left behind (not cleaned up) until the next time I run a snapshot.

**Comment by PhracturedBlue** [comment]
This issue should be closed soon by the devs, as the implementation is complete and released now.

The issue you are reporting in that thread appears to me to be a different bug (though I am not familiar with every part of the snapshot process).  I would recommend filing an independent bug, as it does not appears you have a interval policy defined (which would be the primary reason for getting incomplete snapshots), and so I would not expect your long running snapshots to result in incomplete snapshots.  There are certainly folks who know this system a lot better than I do, but the behavior you are currently seeing does not seem likely to be the desired one.

**Comment by TheRotag** [comment]
Created issue #3347 ("Long-running snapshot leave behind incomplete snapshots and shows 0 byte size")
